# U.S. Medical Insurance Costs

In this project, I will investigate a CSV file with medical insurance costs

To start, I import any libraries I need. For this project the only library I will need is the `CSV` library towork with the **insurance.csv** data. 

In [1]:
# import csv library
import csv

Next, I decide to look through **insurance.csv** and acquaint myself with the data. Things I look for are the following
 - The names of columns and rows
 - Any noticeable missing data
 - Types of values (numerical vs. categorical)

In **insurance.csv*, I saw the following columns
- Patient Age
- Patient Sex
- Patient BMI
- Patient Number of Children
- Patient Smoking Status
- Patient U.S Geopraphical Region
- Patient Yearly Medical Insurance Cost

I did not see any signs of missing data. From this information, I decided to make seven empty lists that will holdeach individual column of data from **insurance.csv**

In [3]:
# In my Python file I create empty lists for the various attributes in insurances.csv

ages = []
sexes = []
bmis = []
num_children = []
smoker_statuses = []
regions = []
insurance_charges = []

I decide to build this function tomake loading data into my lasts as efficient as possible. Without this function I would have to open the **insurance.csv** and rewrite the for loop seven times. However, with this function Ican simply call `load_list_data()` each time as shown below.

In [6]:
# helper function toload csv data
def load_list_data(lst, csv_file, column_name):
    # open csv file
    with open(csv_file) as csv_info:
        # read the data from the csv_file
        csv_dict = csv.DictReader(csv_info)
        # loop through the data in each row of the csv
        for row in csv_dict:
            # add the data from each row to a list
            lst.append(row[column_name])
            
        # return the list 
        return lst

In [56]:
# look at the data in insurance_csv_dict
#load_list_data(ages, 'insurance.csv', 'age')
#load_list_data(sexes, 'insurance.csv', 'sex')
#load_list_data(bmis, 'insurance.csv', 'bmi')
#load_list_data(num_children, 'insurance.csv', 'children')
#load_list_data(smoker_statuses, 'insurance.csv', 'smoker')
#load_list_data(regions, 'insurance.csv', 'region')
#load_list_data(insurance_charges, 'insurance.csv', 'charges')

Now that I have all the data from **insurance.csv** neatly organized into labeled lists, I am ready to start my analysis
- average age of the patients
- number of men vs. women counted in the datadet
- geographical location of the patints
- average yearly medical charges of the patients
- creating a dictionary that contains all patient information

 To perform these inspections, I have built out a class called `PatientsInfo` which contains fives methods
 
 - `analyze_ages()`
 - `analyze_sexes()`
 - `unique_regions()`
 - `average_charges()`
 - `create_dictionary()`

In [35]:
class PatientsInfo:
    # init method that takes in each list parameter
    def __init__(self, patients_ages, patients_sexes, patients_bmis, patients_num_children, 
                 patients_smoker_statuses, patients_regions, patients_charges):
        self.patients_ages = patients_ages
        self.patients_sexes = patients_sexes
        self.patients_bmis = patients_bmis
        self.patients_num_children = patients_num_children
        self.patients_smoker_statuses = patients_smoker_statuses
        self.patients_regions = patients_regions
        self.patients_charges = patients_charges

    # method that calculates the average ages of the patients in insurance.csv
    def analyze_ages(self):
        # initialize total age at zero
        total_age = 0
        # iterate through all ages in the ages list
        for age in self.patients_ages:
            # sum of the total age
            total_age += int(age)
        # return total age divided by the length of the patient list
        return ("Average Patient Age: " + str(round(total_age/len(self.patients_ages), 2)) + " years")

    # method that calculates the number of males and females in insurance.csv
    def analyze_sexes(self):
        # initialize number of males and females to zero
        females = 0
        males = 0
        # iterate through each sex in the sexes list
        for sex in self.patients_sexes:
            # if female add to female variable
            if sex == 'female':
                females += 1
            # if male add to male variable
            elif sex == 'male':
                males += 1
        # print out the number of each
        print("Count for female: ", females)
        print("Count for male: ", males)

    # method to find each unique region patients are from
    def unique_regions(self):
        # initialize empty list
        unique_regions = []
        # iterate through each region in regions list
        for region in self.patients_regions:
            # if the region is not already in the unique regions list
            # then add it to the unique regions list
            if region not in unique_regions: 
                unique_regions.append(region)
        # return unique regions list
        return unique_regions

    # method to find average yearly medical charges for patients in insurance.csv
    def average_charges(self):
        # initialize total_charges variable
        total_charges = 0
        # iterate through charges in patients charges list
        # add each charge to total_charge
        for charge in self.patients_charges:
            total_charges += float(charge)
        # return the average charges rounded to the hundredths place
        return ("Average Yearly Medical Insurance Charges: " +  
                str(round(total_charges/len(self.patients_charges), 2)) + " dollars.")
    
    
    # method to create dictionary with all patients information
    def create_dictionary(self):
        self.patients_dictionary = {}
        self.patients_dictionary["age"] = [int(age) for age in self.patients_ages]
        self.patients_dictionary["sex"] = self.patients_sexes
        self.patients_dictionary["bmi"] = self.patients_bmis
        self.patients_dictionary["children"] = self.patients_num_children
        self.patients_dictionary["smoker"] = self.patients_smoker_statuses
        self.patients_dictionary["regions"] = self.patients_regions
        self.patients_dictionary["charges"] = self.patients_charges
        return self.patients_dictionary

The next step is to create an instance of the class called `patient_info.` With this instance, each method can be used to see the results of the analysis

In [37]:
patient_info = PatientsInfo(ages, sexes, bmis, num_children, smoker_statuses, regions, insurance_charges)


In [38]:
patient_info.analyze_ages()

'Average Patient Age: 39.21 years'

The average age of the patients in **insurance.csv** is about 39 years old. This is important to check in order to ensure the data in **insurance.csv** is representative for a broader populations. if it is decided to use the dataset to make inferences about other populations, the data mustbe abundant and broad enough 

A further analysis would have to be done to make sure the **range** and **standard deviation**** of the patient agegroup in **insurance.csv** is indicative of a random sampling of individuals.

In [44]:
patient_info.analyze_sexes()

Count for female:  662
Count for male:  676


The next step of the analysis is to check the balance of males vs. females in **insurance.csv**. Similar to above, it is important to check that this dataset is representative of a broader population of individuals.

In [40]:
patient_info.unique_regions

['southwest', 'southeast', 'northwest', 'northeast']

There are four unique geographical regions in this dataset, and it is important to note that all the patients come from the United States.

In [42]:
patient_info.average_charges()

'Average Yearly Medical Insurance Charges: 13270.42 dollars.'

The average yearly medical insurance charge per individual is 13270 US dollars.

In [57]:
# patient_info.create_dictionary()